### PR TITLE
[Bug 1431139] Optimize Documents view of kuma

### PR DESCRIPTION
**Pull request by safwanrahman**
In the current platform, it seems like many more unneeded SQL queries that kills most of the time!

It can be more optimized, but for now I think small is also benefecial.

**Current Situation:**
image

**After this Patch:**
image

I was thinking about fetching all the translations data of the document using elasticsearch or cache. As making a sql queries just for showing the translation information is a bit tradeoff with the performance. What do you think @jwhitlock?

@jwhitlock @escattone Can you please review?
It partially fiexes https://github.com/mozmeao/infra/issues/677

### CODE REVIEW

File: kuma/wiki/models.py [hunk 1]
```diff
@@ -413,7 +413,9 @@ def get_zone_subnav_html(self):
         """
         src = self.get_zone_subnav_local_html()
         if (not src) and self.nearest_zone:
-            src = self.nearest_zone.document.get_zone_subnav_local_html()
+            document_id = self.nearest_zone.document_id
+            document = Document.admin_objects.only('html', 'rendered_html', 'zone_subnav_local_html').get(id=document_id)
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: Why ``admin_objects`` (which returns deleted documents as well) instead of ``objects``?

File: kuma/wiki/jinja2/wiki/document.html [hunk 2]
```diff
@@ -202,7 +201,7 @@ <h1>{{ document.title }}</h1>
                 {% include 'wiki/includes/kumascript_errors.html' %}
               {% endif %}
 
-              {% if document.current_revision %}
+              {% if document.parent_id and document.current_revision %}
```
Comment (jwhitlock) [inline kuma/wiki/jinja2/wiki/document.html]: This is probably a good optimization. I'll need to check current docs to see if skips the section when it shouldn't.

File: kuma/wiki/models.py [hunk 3]
```diff
@@ -1375,14 +1377,7 @@ def other_translations(self):
...
         """
-        if self.parent is None:
-            return list(self.translations.all().order_by('locale'))
-        else:
-            translations = list(self.parent.translations
-                                .exclude(id=self.id)
-                                .order_by('locale'))
-            # The parent doc should be at first
-            return [self.parent] + translations
+        return self.get_other_translations()
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: Should ``other_translations`` be deleted instead? If there are other callers, you've made them much less efficient...

File: kuma/wiki/models.py [hunk 4]
```diff
@@ -1466,7 +1474,7 @@ def nearest_zone(self):
     @cached_property
     def is_zone_root(self):
         zone = self.nearest_zone
-        return bool(zone) and (zone.document in (self, self.parent))
+        return bool(zone) and (zone.document_id in (self.id, self.parent_id))
```
Comment (jwhitlock) [inline kuma/wiki/models.py]: 👍 if this removes a query

File: kuma/wiki/utils.py [hunk 5]
```diff
@@ -182,3 +183,12 @@ def analytics_upageviews_by_revisions(revisions):
     start_date = min(r.created for r in revisions)
 
     return analytics_upageviews(revision_ids, start_date)
+
+
+def get_document_tags(document_id):
+    """Get the translations of the document and return locale, slug, title"""
```
Comment (jwhitlock) [inline kuma/wiki/utils.py]: The docstring says it gets translations of the document, but I don't see it in the code. Maybe I need to check the ES return.

File: kuma/wiki/views/document.py [hunk 6]
```diff
@@ -117,8 +117,8 @@ def _get_seo_parent_title(slug_dict, document_locale):
     """
     if slug_dict['seo_root']:
         try:
-            seo_root_doc = Document.objects.get(locale=document_locale,
-                                                slug=slug_dict['seo_root'])
+            seo_root_doc = Document.objects.only('title').get(locale=document_locale,
```
Comment (jwhitlock) [inline kuma/wiki/views/document.py]: 👍 should be kinder to the database

**Comment by metadave** [comment]
@jwhitlock nothing in the [pricing doc](https://www.elastic.co/cloud/as-a-service/pricing) mentions number of requests. As far as I know we pay for high-level infra costs: primary region, # of DC's to host in, CPU/disk size, Google vs AWS.

File: kuma/wiki/utils.py (hunk 5 above)
Comment (safwanrahman) [reply kuma/wiki/utils.py]: Sorry, my bad.
I wrote this docstring for other function, but it got removed later. I will fix it in next commit

File: kuma/wiki/models.py (hunk 3 above)
Comment (safwanrahman) [reply kuma/wiki/models.py]: There are some other callers that can use `get_other_translations()` in order to be more optimized. It makes them much less efficient but changing that can make this PR more complex to review. I can work on that in different PR if you agree.

File: kuma/wiki/models.py (hunk 1 above)
Comment (safwanrahman) [reply kuma/wiki/models.py]: For some [historical purpose](https://bugzilla.mozilla.org/show_bug.cgi?id=1365960), we have seen some error in `DocumentNearestZoneJob` about already deleted document makes `500` error
Therefore used `admin_objects` instead.

**Comment by safwanrahman** [comment]
@jwhitlock Removed the elasticsearch dependency and made more optmization.
Can you give another round of review?

**Comment by jwhitlock** [comment]
@safwanrahman I'll take a look when the tests are passing

**Comment by jwhitlock** [comment]
Thanks @safwanrahman! I'll review on Monday.

**Comment by safwanrahman** [comment]
Thanks @jwhitlock. I need to remove the 6 years old test case and write some other small test cases that cover the full tags functionality. The test case was testing multiple things and calling get request to same endpoint multiple times and because of page cache, the invalid content were served and made the test broken.
I have split up the test in different test cases and place them to the appropiate place where they should be.

**Comment by jwhitlock** [comment]
My process for working with document functionality:

- Fix the code
- Identify the tests that exercise the code and are now broken
- ``git stash`` my code or put it in a local branch
- Refactor the tests to look like I wish they did when I started
- Write a new test that demonstrates the thing that I wanted to fix, to show that it is broken
- Bring in my new code to show I fixed it, without breaking any other code

This results in two commits, a test refactor plus a change in functionality.  See PR #4621 for a recent example.

**Comment by jwhitlock** [comment]
Sorry @safwanrahman, I had too many urgent tasks come up today, it will be tomorrow morning (Tuesday) when I can look at this.

**Comment by jwhitlock** [comment]
There's now a conflict in ``kuma/wiki/jinja2/wiki/includes/buttons.html`` - can you rebase and resolve the conflict?

**Comment by safwanrahman** [comment]
@jwhitlock I have rebased and made a small fix.

**Comment by safwanrahman** [comment]
Thanks @jwhitlock for the review. I will fix it tomorrow.

File: kuma/wiki/jobs.py [hunk 7]
```diff
@@ -127,3 +127,20 @@ def fetch(self, pk, sample_name):
 
     def empty(self):
         return {}
+
+
+class DocumentTagsJob(KumaJob):
```
[the remark this answers is not in the document]
Comment (escattone) [reply kuma/wiki/jobs.py]: @jwhitlock, Great point. We did two things to the `DocumentNearestZoneJob` class to mitigate the spike in jobs:
- Increased the `refresh_timeout` to 180 seconds
- Spread the cache expiration times across a random number of days by overriding the `expiry` method.

For details/reference see https://github.com/mozilla/kuma/pull/4272 and [bugzilla 1373720](https://bugzilla.mozilla.org/show_bug.cgi?id=1373720).

I think we should do the same for `DocumentTagsJob`.

File: kuma/wiki/models.py (hunk 4 above)
Comment (escattone) [reply kuma/wiki/models.py]: @safwanrahman Given your change to the `DocumentNearestZoneJob.fetch` above, which populates the zone's related document, it seems this change is no longer helpful to save a query?

File: kuma/wiki/templatetags/jinja_helpers.py [hunk 8]
```diff
@@ -207,7 +207,7 @@ def document_zone_management_links(user, document):
 
     # Enable "add" link if there is no zone for this document, or if there's a
     # zone but the document is not itself the root (ie. to add sub-zones).
-    if ((not zone or zone.document != document) and
+    if ((not zone or zone.document_id != document.id) and
```
Comment (escattone) [inline kuma/wiki/templatetags/jinja_helpers.py]: @safwanrahman Again, given your change to the `DocumentNearestZoneJob.fetch` above, which populates the zone's related document, it seems this change is no longer helpful to save a query?

File: kuma/wiki/models.py (hunk 4 above)
Comment (safwanrahman) [reply kuma/wiki/models.py]: I think, `self.parent` would make another call if not compare by id!

Comment (escattone) [reply kuma/wiki/models.py]: Ah yes, thanks @safwanrahman!

File: kuma/wiki/templatetags/jinja_helpers.py (hunk 8 above)
Comment (safwanrahman) [reply kuma/wiki/templatetags/jinja_helpers.py]: Yes. it seems to be not saving a query anymore. but I thought abou deferred object can give wrong result in comparing, but it seems that they may make another query if value is not present.
So I think comparing by id would be feasible.

Comment (escattone) [reply kuma/wiki/templatetags/jinja_helpers.py]: @safwanrahman Yes, thanks.

File: kuma/wiki/tests/test_views.py [hunk 9]
```diff
@@ -1681,65 +1682,61 @@ def test_translation_source(self):
...
-                response = self.client.get(reverse('wiki.feeds.recent_documents',
-                                                   args=['atom', t]))
-                self.assertNotContains(response, doc.title)
-
-        # Create a new doc with tags
-        data.update({'slug': slug, 'tags': ','.join(ts1)})
-        self.client.post(reverse('wiki.create'), data)
-        assert_tag_state(ts1, ts2)
+    @pytest.mark.tags
+    def test_tags_showing_correctly_after_doc_update(self):
```
[the remark this answers is not in the document]
Comment (safwanrahman) [reply kuma/wiki/tests/test_views.py]: `test_tags_show_in_document` and this are not same actually.
I wanted to test that after updating the doc with new tag from http request, the page shows the new tag.
Its a edge case that I thought while changing this.

**Comment by safwanrahman** [comment]
@jwhitlock Updated as per commented.
r?

File: kuma/wiki/jobs.py [hunk 10]
```diff
@@ -127,3 +127,27 @@ def fetch(self, pk, sample_name):
...
+    """
+    Given a wiki document returns a list of tags.
+
+    We invalidate this when a document is saved only.
+    Longer lifetime as tags are rarely modified
+    """
+    refresh_timeout = 180
+
+    @property
+    def lifetime(self):
```
Comment (jwhitlock) [inline kuma/wiki/jobs.py]: Not what we asked, but it seems to work.

**Comment by jwhitlock** [comment]
Rebased to merge into a 1) test refactor, and then 2) optimization
